### DISCUSSION TITLE
chapter_backtracking/subset_sum_problem/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#559] [Announcements]
# chapter_backtracking/subset_sum_problem/

在动画与代码中掌握数据结构与算法

https://www.hello-algo.com/chapter_backtracking/subset_sum_problem/

**Comment by zhuasdfg** [comment]
实在是太有石粒辣，感觉这种问题的难点在于剪枝操作怎么写，遍历递归前需要提前考虑的东西很多

**Reply by CXW20001** [reply]
我突然想到，最后一题，如果要求元素只能被选择一次，既然可以排序，那何不直接先去重，再回溯，岂不更秒。哈哈哈

**Reply by krahets** [reply]
@CXW20001 不可以哦，可以试一下看看有什么区别

**Comment by DXShelley** [comment]
根本不是人能想出来的，一环套一环，四环又四环，四环外边是五环。
《新五环之歌》》

**Comment by naivewolf-git** [comment]
重复元素的情况，剪枝四的 i > start 是不是应该改为 i > 0 啊，不然无法剪枝n个元素之和等于目标值的情况。
测试用例：
输入: nums={3,3,4,3,3,5,5,4,5}; target=12;
输出: [[3, 3, 3, 3], [3, 4, 5]]

还有个疑问是可以排序时去重代替剪枝四吗？

**Reply by krahets** [reply]
Hi，这个答案看起来是正确的，为什么需要剪枝“n 个元素之和等于目标值的情况”呢？（每一个 3 都是“不同”的元素）

题目要求“每个元素只能被选择一次”，而非“每种值的元素只能被选择一次”。如果是后者，则可以在数组排序时去重。

**Reply by naivewolf-git** [reply]
概念混淆了，感谢

**Reply by caticat** [reply]
`i > start`和`i > 0`的问题还是不太理解.
为什么这里可以用`start`呢?
如果`start`位置和`start-1`的值是一样的情况,会不会有问题?

**Reply by chenghui-ch** [reply]
start就相当于每次循环的0吧。
这个题不就是判断start和start-1值一样就跳过吗，为啥会有问题🤣

**Comment by muzishuiji** [comment]
subset_sum_ii.js
这部分代码里的
```
// 进行下一轮选择
backtrack(state, target - choices[i], choices, i + 1, res);
```
需要调整为： i+1 -->  i ，不然不满足每个元素可被选取多次的条件～

**Reply by krahets** [reply]
这道题规定“每个元素只可被选择一次”。

**Comment by Mingquan-cloud** [comment]
请教，13.3.2中->“实现方式比较巧妙：由于数组是已排序的，因此相等元素都是相邻的”，题目没有说是给定有序数组nums啊？为什么默认就以有序数组来处理了呢？

**Reply by Mingquan-cloud** [reply]
好吧，看到这段了，偷偷对nums进行排序了。明白了
```cpp
/* 求解子集和 II */
vector<vector<int>> subsetSumII(vector<int> &nums, int target) {
    vector<int> state;              // 状态（子集）
    sort(nums.begin(), nums.end()); // 对 nums 进行排序
```

**Comment by pythonberg1997** [comment]
关于重复情况的剪枝4，其实跟全排列问题中的 `duplicated` 哈希表起到了相同的作用吧，所以代码也可以这样写：
```python
def backtrack(
    state: list[int], target: int, choices: list[int], start: int, res: list[list[int]]
):
    """回溯算法：子集和 II"""
    # 子集和等于 target 时，记录解
    if target == 0:
        res.append(list(state))
        return
    # 遍历所有选择
    # 剪枝二：从 start 开始遍历，避免生成重复子集
    # 剪枝三：从 start 开始遍历，避免重复选择同一元素
    duplicated = set[int]()
    for i in range(start, len(choices)):
        # 剪枝一：若子集和超过 target ，则直接结束循环
        # 这是因为数组已排序，后边元素更大，子集和一定超过 target
        if target - choices[i] < 0:
            break
        # 剪枝四：如果该元素重复，跳过
        if choices[i] in duplicated:
            break
        # 尝试：做出选择，更新 target, start
        state.append(choices[i])
        # 进行下一轮选择
        backtrack(state, target - choices[i], choices, i + 1, res)
        # 回退：撤销选择，恢复到之前的状态
        state.pop()
```

这样可以省去对数组进行排序。我理解的对吗？

**Reply by pythonberg1997** [reply]
漏了 `duplicated.add(choices[i])`

**Reply by pythonberg1997** [reply]
啊好像不行，比那种全排列的情况更复杂

**Reply by ai3027** [reply]
按道理是可以的吧，相当于就是剪掉和他一样的兄弟节点嘛，例题里面是排了序的，可以用那种方式，如果没排序，不就应该用你说这种方式么。

**Reply by zqzwj** [reply]
重复元素应该是continue而不是break，数组中重复元素后面的其它数还能继续选

**Comment by ttppss** [comment]
考虑全排列的情况下，我这种改法是否会有问题？和原始的代码只有一个 slicing 的区别。
```
def backtrack(state, target, choices, res):
    if sum(state) == target:
        res.append(list(state))
        return

    for i, choice in enumerate(choices):
        if sum(state) < target:
            state.append(choice)
            backtrack(state, target, choices[i:], res)
            state.pop()
```

**Reply by krahets** [reply]
用 slicing 也可以的，这是 Python 这类语言的语法糖优势。不过 slicing 会额外创建一个新数组

**Comment by likz1994** [comment]
个人愚见：
13.3.1题目：“给定数组无重复元素，每个元素可以被选取多次”；

13.3.2题目：“给定数组包含重复元素，每个元素只可被选择一次”；

是否中间在增加一个案例：
13.3.X题目：“给定数组无重复元素，每个元素只可被选择一次”；
这样 就每次控制一个 变化量，这样更加 方便理解。这要求是不是有点过分啦  哈哈哈哈

**Reply by hfrangon** [reply]
把中间调用的 backtrack(state, target - choices[i], choices, i, res); 这句中传入i+1就好啦

**Comment by a417197457** [comment]
13.3.2   考虑重复元素的情况

**Reply by a417197457** [reply]
给定数组可能包含重复元素，每个元素只可被选择一次。
如果4,4,5 target为8。 结果应该是4,4
但是上面的解法跳过了4,4 这种可能？

**Reply by krahets** [reply]
不会跳过这种可能呀，测了一下：

```shell
输入数组 nums = [4, 4, 5], target = 8
所有和等于 8 的子集 res = [[4, 4]]
```

建议前往仓库下载代码，并在本地运行。https://github.com/krahets/hello-algo

**Comment by songshengping** [comment]
老板,出书了嘛??有计划嘛??

**Comment by songshengping** [comment]
backtrack(state, target - choices[i], choices, i , res);
我把这行的 i + 1 写成 i 也没有问题.因为我就是想可以重复选择和使用.不知道这样改对不对

**Comment by songshengping** [comment]
建议大佬 把相似的leetcode题目附上,让大家去练

**Reply by krahets** [reply]
Hi，建议直接在题单上找对应分类练习：https://github.com/krahets/LeetCode-Book

**Comment by qw124341789** [comment]
感觉回溯是枚举的另一实现

**Reply by lovelock** [reply]
前面说了回溯就是一种穷举法，所以其实效率不高，只能通过剪枝来尽量提升效率。

**Comment by 2022241104** [comment]
第二项代码的state.pop()换成break会更好一点吗

**Reply by 2022241104** [reply]
抱歉抱歉，我想错了

**Comment by iMissZzz** [comment]
为什么13.3.2中剪枝过程能不像上一份一样用break而改用continue？

**Reply by krahets** [reply]
```py
        # 剪枝四：如果该元素与左边元素相等，说明该搜索分支重复，直接跳过
        if i > start and choices[i] == choices[i - 1]:
            continue
```

Hi，当该剪枝触发时，右边可能还有非重复的搜索分支，因此不能直接 break 跳出，应 continue 继续 for-loop

**Comment by wxnzb** [comment]
为什么要i>start，就跳过

**Reply by Coolllh** [reply]
这是为了防止i=0,i-1数组越界的

**Reply by hfrangon** [reply]
并没有哦 不会i >start就跳过呀 &&为1 还要考虑后面表达式的真值 所以是这一轮中除了第一个元素以外 如果后面还有跟start位置处的元素相同的就应该跳过(在同一轮中不能重复选择)

**Reply by DaiBI5** [reply]
确实，这个i>start表示在本轮选择中的重复元素，如果没有这个限制，一个是会出现： -1下标越界的情况，第二个是会出现： 本轮选择的第一个元素如果与上一轮选择的元素相等时也会被错误的跳过。例如nums=[1,1,2] target=2时，结果集合中不会出现[1,1]这个答案。

**Comment by Coolllh** [comment]
我有个建议,就是在subset_sum_ii.java这个文件里的
```cpp
        // 进行下一轮选择
        backtrack(state, target - choices[i], choices, i + 1, res);
```
加上一个注解,表明这里实际上是剪枝3,把剪枝3标在for循环头上我愣了一会,感觉好像剪枝2和剪枝3的作用是一样的,后面仔细看了代码发现,剪枝3实际上是在递归中体现的

**Reply by krahets** [reply]
谢谢建议！标注在 for 上是因为实际是通过限制 for 循环的范围来实现剪枝的（避免了一部分的选择分支）。在递归调用时，我们给 `start` 参数传入 `i+1` ，给定下一轮的剪枝范围

**Comment by CXW20001** [comment]
我突然想到，最后一题，如果要求元素只能被选择一次，既然可以排序，那何不直接先去重，再回溯，岂不更秒。哈哈哈

**Reply by bunnyHouse** [reply]
假如target是8而不是9，先去重了可如何是好

**Reply by sunshinesDL** [reply]
题目中的“每个元素只可被选择一次”，应该是指每个个体只能用一次，例如，如果有两个值为 1 的元素，它们是两个个体，最后答案中可以出现两个 1 。

**Comment by Jinjin680** [comment]
大佬把回溯讲得好清楚！！！

**Comment by 480284856** [comment]
图13-11:
三哥四哥五哥相邀开黑，但只能两两一起玩，于是三哥首先一个人玩两个角色，很快发现没法玩，于是叫上四哥玩了一局，后面又叫上五哥玩了一局，可是都输了；等到了四哥当房主时，因为觉得三哥是个坑比，四哥直接不和他玩了，于是先自己玩了一局，也发现没法玩，于是叫上五哥玩了一局；五哥发现，四哥也是个坑比，于是后面五哥自己开了个房间，不带三哥和四哥玩了，自己手脚并用都比这两个坑比强。

**Reply by 480284856** [reply]
如图13-10，我们假设没有target的限制，那么这个函数将无限递归下去。首先，每个灰色节点都是一个递归的函数，比如backtrack([], 9, 0, [3,4,5], [])对应根节点（假设函数中的9没有用）,backtrack([], 9, 3, [3,4,5], []), backtrack([], 9, 4, [3,4,5], []), backtrack([], 9, 5, [3,4,5], [])分别对应第二层的三个节点. 如果我们不对choice加以限制，那么第三层的第二个和第四个灰色节点所产生的序列或total的值，都是一模一样的，因为我们可以用[3,4,x,x,...x]和[4,3,x,x,...x]来表示这两个灰色节点下的所有序列。所以，对于父函数选择某个choice，递归到当前函数时（根节点，选择4，递归到第二层的第二个节点），当前节点不能选择比这个choice的值小的choice2（不能选择3），因为父函数在选择choice2，并递归下去时，那个递归函数（第二层的第一个节点）已经选择了这个choice。

**Comment by 480284856** [comment]
在图 13-14中，感觉第三层的第一个节点不应该出现，因为在第二轮选择中，-4^应该被剪枝4去掉。

**Comment by JianhCode** [comment]
写了个自认为比较清晰的代码，如有错误，还请指正。
```cpp
void backtrack(vector<int> &nums, vector<int> &current, int total, int target, vector<vector<int>> &result, int index)
{
    if (total == target)
    { // 如果当前集合之和等于目标，则将当前集合加入结果集
        result.push_back(current);
        return;
    }

for (int i = 0; i < nums.size(); ++i)
{
    if (nums[i] + total > target || i < index || nums[i] == nums[i - 1]) continue;
        // 如果超出边界、出现重复遍历以及重复元素，则跳过
    current.push_back(nums[i]);                                   // 将当前元素加入集合
    backtrack(nums, current, total + nums[i], target, result, i); // 递归调用下一层
    current.pop_back();                                           // 回溯到上一层，移除当前元素
}
}
```

**Reply by WUADN** [reply]
剪枝中i < index有问题，应为 i <= index；因为要不选择重复元素，要从下一个开始遍历，或者backtrack的参数index=i+1；再者，要满足同一轮不重复选择，也应为i + 1。

**Comment by moonpoet** [comment]
重复元素的情况中4种剪枝确实有点复杂，尤其是剪枝二和剪枝三很容易混淆，这两者的维度是不同的，剪枝二是最外层遍历时会逐步减少一个元素，而剪枝三则是往内层遍历时逐步减少一个元素，感觉好像抓住了重点，但是又感觉自己理解的不够深刻，还是需要多看多调试

**Reply by Rumiple** [reply]
剪枝二是backtrack(state, target - choices[i], choices, kk, res);中的kk≥i时成立，这样保证不会遍历小于自身的元素
剪枝三是backtrack(state, target - choices[i], choices, kk, res);中的kk≥i+1时成立，这样保证只会遍历比自身大的元素

**Comment by Maco1010** [comment]
和装载问题很像

**Comment by uwanny** [comment]
我的理解是这样的：
对于第一个问题：
已知原始的数组没有重复的元素，要求列表中不能有重复的组合并且可以选重复的元素，所以下一层要从索引大于等于上一层选择的位置选择。
对于第二个问题：
已知列表有重复的元素，要求列表中不能有重复的组合并且不能选重复的元素，所以下一层要从索引大于上一层选择的位置选择。但是只这样要求依然会出现重复子集的情况，比如作者给出的例子【4，5】【不同的4，5】，造成这种情况的原因是数组有重复的元素（细节我也不太清楚），所以我们要限制在同一层下一个元素不能和上一个一样，但是下一层的元素可以和上一层的一个样，只要是不同的索引。

**Comment by JameRoyar** [comment]
其实很简单 剪枝二和剪枝三 第一次看注释为什么都是start开始遍历还有两个结果的 瞬间云里雾里 
其实一个是从i开始的 一个是从i+1开始的 这里是一个没有说明白的地方很容易造成误导 注释的意义应该是让代码更清晰而不是相反的
从i开始保证了不会倒回去和原来算过的元素再结合一遍
从i+1开始是把自己和自己的结合的情况剪枝掉了

**Reply by syutengu** [reply]
是的，i是下一轮选择中保留nums[i]再次被选的可能，i+1就是把nums[i]自身也一起剪掉。

**Comment by LJQ0** [comment]
可否在这请教各位佬一道算法题，之前做的没做上来，学到这想想有点像，又有点不一样
https://acexam.nowcoder.com/coding/?uid=54AB4E4B3B08F3DD&qid=10925271

**Reply by wdllmh** [reply]
应该不是，这道题如果用每个位置的回溯，感觉可能性太多了，应该是根据按位或的性质去生成最有可能的序列进行判断

**Comment by MambaFly** [comment]
相等元素既有自身和自身相等，又有排序后相邻相等元素，因此分别对应剪枝3和剪枝4

**Comment by BLackWatch1118** [comment]
简单分享下自己的理解：
不要把这个问题按照数学上的思路理解为对元素的选择，因为代码中所有的选择都是从choices数组中选取，应当把它理解为对位置的选择，也就是对`[0, 1, ~, n-1]`索引子集的选择。在第一个问题中重复产生的原因是所选索引子集的大小顺序不同的，所选择的索引都是一样的，所以只需要保证选择的索引为正序，就能够排除重复子集的情况。

对于第二个问题，由于已将数组排序，大小相同的元素会排列在一起，在每轮选择排除重复元素的方法是判断此选择是否与前一个元素相同，注意剪枝4判断中会有`i > start `的条件是因为start是循环中选择的首个元素，它的前一个元素在上一轮中被选择过不违反每轮中不能选择重复元素的要求。每个元素只能选择一次通过限定每次调用递归函数时令start+1实现，这个就是在前一个问题下增加了下一轮不能选择自身的条件。

**Comment by ruihao-yan** [comment]
昨天晚上梦见宿管阿姨问我回溯算法，我全说上来了:D

**Comment by Asce-ops** [comment]
重复子集那里感觉解释得不太清楚，我理解应该是当做出选择 𝑥𝑖 后，设定下一轮从索引 𝑖 + 1 开始向后遍历和相等元素在每一轮中只能被选择一次共同保证不会出现重复子集

**Comment by Arman19941113** [comment]
感觉重复元素的情况解法和题意应该出现了偏差？题目是说每个元素只可被选择一次，题解是相等元素在每一轮中只能被选择一次，这道题应该是对应 LeetCode 组合总和 2 这道题，需要引入一个 visited 辅助数组来解答。

**Reply by gfzum** [reply]
这里的描述确实有些偏差，但是解法依然是 lc 组合总和2 那道题的解法。尽管在排序后，第二次遍历到相等数字会直接跳过，但是第一次遇到该数字的下次 backtrack 不会跳过和他相等的数字，所以可以保证是 “相同元素”只出现一次，“相等元素”还是会被选择到的。

**Reply by 3Ajfdjsdsk2** [reply]
我个人认为没有偏差。每个元素被选择一次是指不同轮里面，也就是结果数组中不同索引不能选取source里面同一位置元素。而相等元素在每一轮中只能选择一次指的是结果数组中同一个索引对应的元素

**Reply by Arman19941113** [reply]
嗯，没啥问题，我看岔了😂

**Comment by syutengu** [comment]
学习笔记，js版
```js
  /* k数之和，含重复元素，回溯法，可选择是否允许元素复用 */
  function k_sum_backtracking(nums, k, can_reuse = false) {
    const res = []
    const sub = []
    backtrack()
    return res

    function backtrack(j = 0) {//记录前一轮递归中指针i到达的位置，以实现剪枝不同选择顺序造成的重复子集
      const vals = new Set()//记录使用过的值，以实现剪枝元素同值造成的重复子集
      if (!k) return res.push([...sub])

      for (let i = j; i < nums.length; i++) {
        const num = nums[i]
        if (k - num < 0 || vals.has(num)) continue

        /* 进 */
        sub.push(num)
        k -= num
        vals.add(num)
        backtrack(i + !can_reuse)//后一轮递归从i（不允许复用元素时，i+1）位置开始遍历元素
        /* 退 */
        sub.pop(num)
        k += num
      }
    }
  }
```

**Reply by syutengu** [reply]
应该是n数之和k

**Reply by syutengu** [reply]
递归也应该用”层“而不是”轮“作为量词计数比较好吧。

**Reply by syutengu** [reply]
sub.pop(num) ❌
sub.pop() ✔

**Reply by syutengu** [reply]
和lc#18的要求搞混了，这个没有元素个数限制，所以名字就应该叫子集和subset_sum。

**Comment by huanghejian** [comment]
关于重复哪个，为什么不直接数组去重

**Reply by wdllmh** [reply]
因为直接去重的话会导致需要多个相同元素的解不能被搜到，比如[2,2,3]中找4，就需要2+2

**Comment by YiChau** [comment]
LeetCode 对应题目： 1. 无重复元素：https://leetcode.cn/problems/combination-sum/description/
2. 考虑重复元素https://leetcode.cn/problems/combination-sum-ii/

**Comment by Weining2020** [comment]
Hi, 
对于“13.3.2   考虑重复元素的情况 2.代码实现” 部分，我没找到剪枝二的实现。我看的是C#的代码。我看到注释 “剪枝二：从 start 开始遍历，避免生成重复子集” 位置还保持在for循环之上。我明白对于无重复元素的情况，剪枝二是由i=start实现的，可是对于有重复元素的情况，即使i=start也不能避免 choices[start]和choices[start - 1]是相等元素值的情况吧？也就是不能避免生成重复子集？还是说 if (i > start && choices[i] == choices[i - 1]) 代码同时做到了剪枝二和剪枝四的实现？

**Reply by jiuerya7** [reply]
是i = start 和递归调用设置成 i + 1同时满足了剪枝2和剪枝3，如果递归里面填入的是i，仅满足剪枝2：就是避免结果序列有重复的结果

**Comment by Tianle-Leviathan** [comment]
对于“13.3.2 考虑重复元素的情况 2.代码实现” 部分，有没有一种解法是不需要nums数组有序列的？有序的话可以保证重复数字是连续出现的，所以可以很方便的在同一个函数调用的frame里剪枝。如果nums是无序的，那么很有可能出现类似[1,2]和[2,1]这类的重复。

**Reply by lizhen921** [reply]
不排序的话，需要利用duplicated map去重，应该也是可以的吧

**Reply by jiuerya7** [reply]
用哈希表试试

**Reply by jiuerya7** [reply]
我试了试，不太行，还是先把数组排序再回溯吧，直接写针对无序有重复的输入，情况很复杂，需要将state排序，然后借助一个额外的哈希表，逐个对比哈希表，有没有重复的，然后再添加到res，很麻烦，时间复杂度和空间复杂度还不如这个，还是算了

**Reply by jiuerya7** [reply]
先排序再使用有序且输入可重复的算法，比直接使用 无序且可重复的算法更节省时间和空间

**Comment by xllzi** [comment]
考虑重复元素的组合问题的算法时间复杂度应该为O(target的拆分数)，有兴趣的朋友可以学习一下相关数论知识:)

**Comment by ergouHZ** [comment]
```cpp
// 剪枝三：从 start 开始遍历，避免重复选择同一元素

```
这个地方应该标在下一轮递归 start 使用了i+1更清晰，看了半天搞明白了
```cpp
backtrack(state, target - choices[i], choices, i + 1, res);
```

**Comment by HansSeven** [comment]
这里我有个想法，既然是要防止重复，我想的是
1.首先，必须要不重复选自身之前的数字（自身可多选）
2.出现的问题本质上是结果的无序导致重复，如果有序，本身是不会选出同样的结果的
核心问题是结果的有序性，（start标记本事是索引的有序性）
3.因此可以对原数组排序，或者在选择choice的时候只选择比该值大或者相等（自身可复选）的数，这样的话也维持了结果的有序性，即不会选出结果一样的序列。
```python
def backtrack(state:list[int],res:list[list[int]],target:int,choices:list[int],total):
    if total ==target :
        res.append(state.copy())
    for choice in choices:
        if (total+choice)<target and choice>=state[-1]:
            state.append(choice)
            total+=choice
            backtrack(state,res,target,choices,total)
            total-=choice
            state.pop()
```
有重复元素的情况只需要加一个set（）判断是否出过该结果来剪枝就行

**Comment by CYX22222003** [comment]
感觉Comb Sum 2关键的invariant是，在下一轮的选择必须大于等于上一轮的选择
